### ISSUE TITLE
[rlgl] WebGL: texImage2D: invalid internalformat

### ISSUE THREAD

**Issue by BugraAlptekinSari** [raysan5/raylib#2824]
I am trying to create a FBO with a depth buffer as texture2D in a web build using OpenGL ES2.

to do this, I first create the FBO using a slightly modified version of loadRenderTexture() like this:
```c
RenderTexture2D LoadRenderTextureMODIFIED(int width, int height) {
    //unchanged lines
        target.depth.id = rlLoadTextureDepth(width, height, false);
    //unchanged lines
        rlFramebufferAttach(target.id, target.depth.id, RL_ATTACHMENT_DEPTH, RL_ATTACHMENT_RENDERBUFFER, 0);
    //unchanged lines
}
```
The FBO object created with this single change completely failed to render because web gl threw `incomplete FBO` error after rejecting the internal format of the depth texture.
To overcome this issue, I created a custom rlLoadTextureDepth() function and manually inserted the available internalformats one by one. 
```c
// Removed all safety checks.
// Removed support for anything other than GL ES 2.0
unsigned int rlLoadTextureDepthMODIFIED(int width, int height, bool useRenderBuffer)
{
    unsigned int id = 0;

    // NOTE: We let the implementation to choose the best bit-depth
    // Possible formats: GL_DEPTH_COMPONENT16, GL_DEPTH_COMPONENT24, GL_DEPTH_COMPONENT32 and GL_DEPTH_COMPONENT32F
        unsigned int glInternalFormat = GL_DEPTH_COMPONENT32_OES; //Manually tested every possible value here.

        glGenTextures(1, &id);
        glBindTexture(GL_TEXTURE_2D, id);
        glTexImage2D(GL_TEXTURE_2D, 0, glInternalFormat, width, height, 0, GL_DEPTH_COMPONENT, GL_UNSIGNED_INT, 0);

        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);

        glBindTexture(GL_TEXTURE_2D, 0);

        TRACELOG(RL_LOG_INFO, "TEXTURE: Depth texture loaded successfully");
 
    return id;
}
```
None of that helped so I also tried changing the value format argument of glTexImage2D() function (GL_UNSIGNED_INT , GL_SHORT, GL_FLOAT32 etc.)
That didn't change anything either.
I also tried recompiling raylib a couple times and re-install emsdk several times to make sure I didn't get any of those wrong.

The computer I am working with has Ubuntu 20.4 installed.
I also tried compiling on my windows10 machine to make sure that isn't the cause.

**Comment by raysan5** [comment]
@BugraAlptekinSari An FBO requires at least one color attachment to be valid, on `LoadRenderTextureMODIFIED()` you only attach a depth texture.

**Comment by BugraAlptekinSari** [comment]
@raysan5 This is the actual code
```c
unsigned int rlLoadTextureDepthMOD(int width, int height){
  unsigned int id = 0;

  glGenTextures(1, &id);
  glBindTexture(GL_TEXTURE_2D, id);
  glTexImage2D(GL_TEXTURE_2D, 0, GL_DEPTH_COMPONENT24_OES, width, height, 0, GL_DEPTH_COMPONENT, GL_UNSIGNED_INT, 0);

  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);

  glBindTexture(GL_TEXTURE_2D, 0);

  TRACELOG(RL_LOG_INFO, "TEXTURE: Depth texture loaded successfully");
  return id;
}

RenderTexture2D LoadRenderTextureMOD(int width, int height)
{
    RenderTexture2D target = { 0 };

    target.id = rlLoadFramebuffer(width, height);   // Load an empty framebuffer

    if (target.id > 0)
    {
        rlEnableFramebuffer(target.id);

        // Create color texture (default to RGBA)
        target.texture.id = rlLoadTexture(0 , width, height, PIXELFORMAT_UNCOMPRESSED_R8G8B8A8, 1);
        target.texture.width = width;
        target.texture.height = height;
        target.texture.format = PIXELFORMAT_UNCOMPRESSED_R8G8B8A8;
        target.texture.mipmaps = 1;

        // Create depth renderbuffer/texture
        target.depth.id = rlLoadTextureDepthMOD(width, height);
        target.depth.width = width;
        target.depth.height = height;
        target.depth.format = 19;       //DEPTH_COMPONENT_24BIT?
        target.depth.mipmaps = 1;

        // Attach color texture and depth renderbuffer/texture to FBO
        rlFramebufferAttach(target.id, target.texture.id, RL_ATTACHMENT_COLOR_CHANNEL0, RL_ATTACHMENT_TEXTURE2D, 0);
        rlFramebufferAttach(target.id, target.depth.id, RL_ATTACHMENT_DEPTH, RL_ATTACHMENT_TEXTURE2D, 0);

        // Check if fbo is complete with attachments (valid)
        if (rlFramebufferComplete(target.id)) TRACELOG(LOG_INFO, "FBO: [ID %i] Framebuffer object created successfully", target.id);

        rlDisableFramebuffer();
    }
    else TRACELOG(LOG_WARNING, "FBO: Framebuffer object can not be created");

    return target;
}
```
The comments saying `//unchanged lines` were meant to show the lines other than the ones I wrote are **unchanged**. Meaning I did not remove or add anything in between.
Another thing I would like to stress is that **This code works for anything other than WEB PLATFORM**. The issue is specific to opengl es 2.0

Please spend more than a minute reading these issues. Responses like these are an insult to the effort I spent on this problem.

**Comment by raysan5** [comment]
> Please spend more than a minute reading these issues. Responses like these are an insult to the effort I spent on this problem.

@BugraAlptekinSari I spend more than a minute and I reviewed rlgl implementation. Kind reminder that raylib is an open source project, I'm not paid to solve other people problems or even invest a minute to review them, I do it for free.

**Comment by raysan5** [comment]
Just reviewed this issue more carefully.

I noticed `GL_OES_depth24` and `GL_OES_depth32` extensions are not detected on WebGL and `GL_DEPTH_COMPONENT16` is not understood by WebGL as internal format for `glTexImage2D()`.

raylib OpenGL ES 2.0 is translated by emscripten to WebGL 1.0 but it seems specifications differ between both implementation, OpenGL ES 2.0 supports sized internal formats for depth textures while WebGL does not. 

In any case, the valid internal format for WebGL depth textures is just `GL_DEPTH_COMPONENT`, as noted in [Mozilla official WebGL documentation](https://developer.mozilla.org/en-US/docs/Web/API/WEBGL_depth_texture).

Reviewed the code accordingly.

Also important to note for the future: As per the [official Khronos documentation](https://registry.khronos.org/webgl/extensions/WEBGL_depth_texture), about the extension `WEBGL_depth_texture`:

 > No longer available as of the [WebGL API 2.0](http://www.khronos.org/registry/webgl/specs/2.0/) specification.

Feel free to reopen if problem persists.

**Comment by BugraAlptekinSari** [comment]
Thank you. I think that solves the problem.
Also I would like to apologize for my rash response.

I had assumed you were being dismissive because the solution you initially offered was unrelated to the error message and If it really were the solution then this specific error message would not have appeared at all. However it was presumptuous of me to assume there would be no other kind of mistake.
